### ISSUE TITLE
Convert types to use PhetioObject

### ISSUE THREAD

**Issue by samreid** [phetsims/tandem#46]
See https://github.com/phetsims/phet-io/issues/1227

**Comment by samreid** [comment]
Beer's Law Lab is now using PhetioObject, see https://github.com/phetsims/beers-law-lab/issues/219

**Comment by samreid** [comment]
I've done several, but there are still 30 or so more cases to convert.  Tagging @ariel-phet and @kathy-phet so they are aware I'll be working on this.

**Comment by samreid** [comment]
I've converted to use PhetioObject where possible.  There are still a few singletons using tandem.addInstance, but they look appropriate.  The next step will be to turn on the assertions in PhetioObject, but that will be in another issue.

**Comment by samreid** [comment]
Created new issue #48, closing.

**Comment by zepumph** [comment]
There is one todo that was added in this issue without an issue attached to is, assigning to you @samreid.

In Property.dispose():

```js
 dispose: function() {

        assert && assert( !this.isDisposed, 'cannot be disposed twice' );

        this.isDisposed = true; // TODO: move this to PhetioObject?

        // remove any listeners that are still attached to this property
        this.unlinkAll();

        PhetioObject.prototype.dispose.call( this );
      },
```

**Comment by samreid** [comment]
Node.js declares:

```js
    /**
     * Whether this Node has been disposed.
     * @public
     *
     * @returns {boolean}
     */
    isDisposed: function() {
      return this._isDisposed;
    },
    get disposed() { return this.isDisposed(); },
```

PhetioObject currently has `phetioObjectDisposed`, perhaps so named to avoid collisions with Property and Node.  @jonathanolson what do you think about Node leveraging PhetioObject disposed flag?

**Comment by jonathanolson** [comment]
> @jonathanolson what do you think about Node leveraging PhetioObject disposed flag?

Sounds fine to me. If it's being elevated from phet-io to "general", can it be called _isDisposed (and similarly Property would use it?)

Also make sure that the disposal logic (if assertions are enabled) below the _isDisposed definition will work.

**Comment by samreid** [comment]
@jonathanolson would it be OK if I go with the name `isDisposed` (no underscore) and mark it as `@protected`, and eliminate the method `isDisposed()`?

**Comment by zepumph** [comment]
What about having a `this._isDisposed` that is private, and then have a protected es5 getter and setter `this.isDisposed`?

**Comment by samreid** [comment]
Can you elaborate on the advantage of having underscore + es5 getter+setter as opposed to just a protected `this.isDisposed`?

**Comment by zepumph** [comment]
No strong preference; to me it seems more idiomatic to OO programming, and is also the way that Node is built.

**Comment by jonathanolson** [comment]
> @jonathanolson would it be OK if I go with the name isDisposed (no underscore) and mark it as @protected, and eliminate the method isDisposed()?

I can't really imagine any need to override whether something appears to be disposed, so I don't have a preference.

**Comment by samreid** [comment]
I'll mark this as blocks-publication until it has been tested/reviewed, since this will affect several scenery and axon files and usages.

**Comment by samreid** [comment]
I've made local changes to unify Node._isDisposed and Property.isDisposed with PhetioObject.isDisposed.  I am not 100% confident that the changes are correct (particularly in Node), but the changes are across 13 files in 8 repos, so not trivial to create branches.

Local aqua testing is good so far after about 12 sims.

I will also need to leak test.  I'm testing Graphing Quadratics since it's been recently memory tested, but I don't know that it calls dispose() much.

**Comment by samreid** [comment]
Local aqua is showing no issues and I think I'm at a good point to commit and get review feedback.  @jonathanolson do you have time to take a look?

**Comment by jonathanolson** [comment]
>  Regarding 'Node.dispose() call is missing from an overridden dispose method' -- do we want that functionality it PhetioObject?

Presumably yes. Can test in the Scenery playground:
```js
function SubNode() { scenery.Node.call( this ); }
phetCore.inherit( scenery.Node, SubNode, {
  dispose() { console.log( 'oops no super call' ) }
} );
var s = new SubNode();
s.dispose(); // no error
```

however if it's extremely rare to run into this for all devs, then I'm fine dropping it. Presumably the code could just be moved over to PhetioObject? (I'll test that and see if I run into assertions anywhere).

**Comment by jonathanolson** [comment]
The other commits (tandem/axon/scenery) look good.

**Comment by jonathanolson** [comment]
Adding the assertion found that ESPB/ESP Track.js inherits PhetioObject but doesn't do the super dispose.

**Comment by samreid** [comment]
That sounds like a good catch, do you want to commit the change to PhetioObject?

**Comment by pixelzoom** [comment]
@samreid will commit the fix, then remove the "blocks" label.  Wait on GQ and AM RCs until this is done.

**Comment by samreid** [comment]
Proposed fixes above and correctly flag problems in Energy Skate Park sims.  Flagging @jessegreenberg so he is aware.  Reassigned to @jonathanolson to review and close if all is well.

**Comment by pixelzoom** [comment]
Increasing priority, since this is blocking Graphing Quadratics 1.0 RC. Please notify me when this is closed.

**Comment by samreid** [comment]
I also mentioned to @pixelzoom that he can review this if he gets to it first.

**Comment by pixelzoom** [comment]
I'd prefer not to spend time familiarizing myself with this issue.  But if @jonathanolson doesn't get to it by Monday morning, I'll dive in.

**Comment by jonathanolson** [comment]
Looks good to me.
